plugins/imagebase/plugin.js

```diff
@@ -0,0 +1,113 @@
+/**
+ * @license Copyright (c) 2003-2017, CKSource - Frederico Knabben. All rights reserved.
+ * For licensing, see LICENSE.md or http://ckeditor.com/license
+ */
+
+( function() {
+	'use strict';
+
+	function createWidgetDefinition( editor, definition ) {
+		var defaultTemplate = new CKEDITOR.template(
+			'<figure class="{basicClass}">' +
+				'<img alt="" src="" />' +
+				'<figcaption>{captionPlaceholder}</figcaption>' +
+			'</figure>' ),
+			baseDefinition;
+
+		/**
+		 * This is an abstract class that describes a definition of a basic image widget
+		 * creted by {@link CKEDITOR.plugins.imagebase#addImageWidget} method.
+		 *
+		 * Note that because the image widget is a type of a widget, this definition extends
+		 * {@link CKEDITOR.plugins.widget.definition}.
+		 * It adds several parts of image and implements the basic version of
+		 * {@link CKEDITOR.plugins.widget.definition#upcast} callback.
+		 *
+		 * @abstract
+		 * @since 4.8.0
+		 * @class CKEDITOR.plugins.imagebase.imageWidgetDefinition
+		 * @mixins CKEDITOR.plugins.widget.definition
+		 */
+		baseDefinition = {
+			pathName: editor.lang.imagebase.pathName,
+
+			template: defaultTemplate,
+
+			allowedContent: {
+				img: {
+					attributes: '!src,alt,width,height'
+				},
+				figure: true,
+				figcaption: true
+			},
+
+			requiredContent: 'figure',
+
+			editables: {
+				caption: {
+					selector: 'figcaption',
+					pathName: editor.lang.imagebase.pathNameCaption,
+					allowedContent: 'br em strong sub sup u s; a[!href,target]'
+				}
+			},
+
+			parts: {
+				image: 'img',
+				caption: 'figcaption'
+			},
+
+			upcasts: {
+				figure: function( element ) {
+					// http://dev.ckeditor.com/ticket/11110 Don't initialize on pasted fake objects.
+					if ( element.attributes[ 'data-cke-realelement' ] ) {
+						return;
+					}
+
+					if ( element.name === 'figure' ) {
+						return element;
+					}
+				}
+			}
+		};
+
+		definition = CKEDITOR.tools.object.merge( baseDefinition, definition );
+
+		/**
+		 * Image widget definition overwrites the {@link CKEDITOR.plugins.widget.definition#upcast} property.
```

Comment:
This should be a bit better documented, just briefly what it does? We want to avoid docs like "This is a property that should that should not interest you! Do not touch!" 😉 

How about something simple, like:

`Overrides {@link CKEDITOR.plugins.widget.definition#upcast} property, it's automatically set to enumerate keys of {@link #upcasts}. Avoid changes, unless you know what you're doing.`

?